src/project/internal/recentfilescontroller.cpp

```diff
@@ -0,0 +1,263 @@
+}
+
+void RecentFilesController::removeNonexistentFiles()
+{
+    bool removed = false;
+
+    RecentFilesList newList;
+    newList.reserve(m_recentFilesList.size());
+
+    for (const RecentFile& file : m_recentFilesList) {
+        if (fileSystem()->exists(file)) {
+            newList.push_back(file);
+        } else {
+            removed = true;
+        }
+    }
+
+    if (removed) {
+        setRecentFilesList(newList, false);
+
+        async::Async::call(nullptr, [this, newList]() {
+            saveRecentFilesList();
+        });
+    }
+}
+
+void RecentFilesController::setRecentFilesList(const RecentFilesList& list, bool save)
+{
+    if (m_recentFilesList == list) {
+        return;
+    }
+
+    m_recentFilesList = list;
+
+    cleanUpThumbnailCache(list);
+
+    if (save) {
+        saveRecentFilesList();
+    }
+}
+
+void RecentFilesController::saveRecentFilesList()
+{
+    TRACEFUNC;
+
+    m_reloadingBlocked = true;
+
+    DEFER {
+        m_reloadingBlocked = false;
+    };
+
+    QJsonArray jsonArray;
+    for (const RecentFile& file : m_recentFilesList) {
+        jsonArray << file.toQString();
+    }
+
+    QJsonDocument jsonDoc(jsonArray);
+    QByteArray json = jsonDoc.toJson(QJsonDocument::Compact);
+
+    mi::WriteResourceLockGuard resource_guard(multiInstancesProvider(), RECENT_FILES_RESOURCE_NAME);
+    Ret ret = fileSystem()->writeFile(configuration()->recentFilesJsonPath(), ByteArray::fromQByteArrayNoCopy(json));
+    if (!ret) {
+        LOGE() << "Failed to save recent files list: " << ret.toString();
+    }
+}
+
+Promise<QPixmap> RecentFilesController::thumbnail(const RecentFile& file) const
+{
+    return Promise<QPixmap>([this, file](auto resolve, auto reject) {
+        if (file.empty()) {
+            return reject(int(Ret::Code::UnknownError), "Invalid file specified");
+        }
+
+        QtConcurrent::run([this, file, resolve, reject]() {
+            std::lock_guard lock(m_thumbnailCacheMutex);
+
+            DateTime lastModified = fileSystem()->lastModified(file);
+
+            auto it = m_thumbnailCache.find(file);
+            if (it != m_thumbnailCache.cend()) {
+                if (lastModified == it->second.lastModified) {
+                    (void)resolve(it->second.thumbnail);
+                    return;
+                }
+            }
+
+            RetVal<ProjectMeta> rv = mscMetaReader()->readMeta(file);
+            if (!rv.ret) {
+                m_thumbnailCache[file] = CachedThumbnail();
+                (void)reject(rv.ret.code(), rv.ret.toString());
+            } else {
+                m_thumbnailCache[file] = CachedThumbnail { rv.val.thumbnail, lastModified };
+                (void)resolve(rv.val.thumbnail);
+            }
+        });
+
+        return Promise<QPixmap>::Result::unchecked();
+    }, Promise<QPixmap>::AsynchronyType::ProvidedByBody);
+}
+
+void RecentFilesController::cleanUpThumbnailCache(const RecentFilesList& files)
+{
+    QtConcurrent::run([this, files] {
```

Comment:
Why is QtConcurrent needed here?

Reply:
In order not to block the UI waiting for the mutex. If we would do that on the main thread, then we would first block the UI, waiting for all concurrent threads, until we can acquire the mutex lock, and then we can finally do the operation. 

Of course, when using QtConcurrent, the correct order might not be guaranteed (I can't find any promises about this in the documentation). But that does not matter much:
- if a thumbnail is inserted before cleaning, but it should have happened after, then it means that the thumbnail will have to be recomputed the next time it is requested. Not a big problem. In the majority of the cases, the cache does still do its job properly.
- if a thumbnail is inserted after cleaning, but it should have happened before, then the thumbnail will stay in the cache until the next cleanup. So one thumbnail remains in the cache a little longer. Not a big problem either. The only reason that we are doing these cleanups at all is to reduce useless memory usage, but one thumbnail more or less does not make a big difference.